### ISSUE TITLE
Rrule.after calls are very slow

### ISSUE THREAD

**Issue by dangdennis** [jakubroztocil/rrule#407]
## Reporting an issue
The `.after()` calls are very slow. 

### Version
2.6.4

### Expected Output
I was hoping occurrence time calculation are faster. 200-300ms is too slow. 

#### Benchmarks

I've included a codesandbox for easy viewing of a some benchmarking.
https://codesandbox.io/s/rrulesetafter-benchmark-lnkk5
```
// rruleset 1
compute time building rrule:  1.5642419997602701
compute time calculating next occurrence:  290.0715749990195

// rruleset 2
compute time building rrule:  0.3221629988402128
compute time calculating next occurrence:  265.0948979984969
```

### OS
MacOS 10.15.5

### Date
Thu Jun 18 00:36:51 PDT 2020

### Note

- [X] Verify that you've looked through existing issues for duplicates before
  creating a new one
- [X] Code sample reproducing the issue. Be sure to include all input values you
  are using such as the exact RRule string and dates.
- [X] Expected output
- [X] Actual output
- [X] The version of `rrule` you are using
- [X] Your operating system
- [X] Your local timezone (run `$ date` from the command line
  of the machine showing the bug)

**Comment by msageryd** [comment]
I landed here after having the same problem. The calculations seems to be slower when more passed occurrances exists.

This is easily tested by setting frequency to seconds and set start date to a couple of days ago. Two days worth of seconds (approx 170,000) makes the .after() call take several seconds on my speedy MacbookPro.

If the start date is set to only some minutes ago the calculation is very fast.

```
const test2 = new RRule.fromString(
  'DTSTART;TZID=America/Adak:20200812T111100\nRRULE:FREQ=SECONDLY;WKST=MO'
);

console.log(test2.after(new Date()));
```

**Comment by dangdennis** [comment]
Thanks for sharing. I’ll have to try this, as it may be a good workaround for me. Otherwise, I’m running a nightly batch job to get the latest relevant occurrence.

**Comment by dangdennis** [comment]
I’ll close this since there’s already an existing one.

**Comment by msageryd** [comment]
@dangdennis Check out my function `calcSafeDTstart()` in my soon to launch library: https://github.com/msageryd/node-schedule-rrule#readme

You can find it at the end of the file recurrenceRule.js in the lib folder.

Please don't forget to ping back if you find any problems with the function (feel free to create an issue). All my tests passes, but I should probably create some specialised tests to stress this function as well.

**Comment by dangdennis** [comment]
Wonderful @msageryd! I’ll be happy to try out your function and let you know. Phew, glad you paved the way. I was considering migrating some APIs over to Go just for this perf reason.

**Comment by lorenzodejong** [comment]
Perhaps opening up an old thread, but i ran into a similar issue with high frequency intervals (minutely/secondly) and also needed a mechanism to determine the next occurrence.

@msageryd i like your solution, i've taken some inspiration from it in my own codebase. However one thing i noticed is that it does not always calculate the correct date. For example you're always subtracting 31 days with a monthly frequency. Months obviously have varying amounts of days, you may end up with a different reoccurrence than intended by the RRULE.

I ended up using a date library (like `date-fns`) to subtract the correct amount of time from the current date. It comes with a set of helpers to achieve this:
- [subYears](https://date-fns.org/v4.1.0/docs/subYears)
- [subMonths](https://date-fns.org/v4.1.0/docs/subMonths)
- [subDays](https://date-fns.org/v4.1.0/docs/subDays)
- [subHours](https://date-fns.org/v4.1.0/docs/subHours)
- [subMinutes](https://date-fns.org/v4.1.0/docs/subMinutes)
- [subSeconds](https://date-fns.org/v4.1.0/docs/subSeconds)

**Comment by msageryd** [comment]
@lorenzodejong Thanks for the heads-up. I'll check this out if/when I get time to move forward with my library. My own need for the library was pushed into the future, so I haven't put any more time into this.

I'm glad you found it useful.

**Comment by msageryd** [comment]
@lorenzodejong I got curious and checked out the code. It is actually not important to hit the exact date "one month ago". The most important is that we don't under-shoot, i.e. back off by 30 days in a 31 day month. By backing off 31 days for monthly intervals we cover this.

But I did spot a more serious bug. When backing off the date for yearly intervals I deduct 365 days. This won't work for leap years. I would back off by 366 days instead to make sure.

Since the exactness of this is not important, I would rather not bring in another dependency like date-fns.

---- edit:
I'll have to back off of my above statement. This only holds true if the RRule includes some more parameters which makes dtStart not that important, such as `byweekday: [RRule.MO, RRule.FR]`

Most rules will rely on an exact dtStart to calculate the next occurance, hence date-fns or alike will be needed.
